src/awkward/_backends.py

```diff
@@ -0,0 +1,189 @@
+from __future__ import annotations
+
+from abc import abstractmethod
+
+import awkward_cpp
+
+import awkward as ak
+from awkward._typetracer import NoKernel, TypeTracer
+from awkward.nplikes import (
+    Cupy,
+    CupyKernel,
+    Jax,
+    JaxKernel,
+    Numpy,
+    NumpyKernel,
+    NumpyLike,
+    NumpyMetadata,
+    Singleton,
+    nplike_of,
+)
+from awkward.typing import (
+    Any,
+    Callable,
+    Protocol,
+    Tuple,
+    TypeAlias,
+    TypeVar,
+    Unpack,
+    runtime_checkable,
+)
+
+np = NumpyMetadata.instance()
+
+
+T = TypeVar("T")
+KernelKeyType: TypeAlias = Tuple[str, Unpack[Tuple[np.dtype, ...]]]
+KernelType: TypeAlias = Callable[..., None]
+
+
+@runtime_checkable
+class Backend(Protocol[T]):
+    @property
+    @abstractmethod
+    def nplike(self) -> NumpyLike:
+        ...
+
+    @property
+    @abstractmethod
+    def index_nplike(self) -> NumpyLike:
+        ...
+
+    def __getitem__(self, key: KernelKeyType) -> KernelType:
+        ...
+
+
+class NumpyBackend(Backend[Any], Singleton):
+    _numpy: Numpy
+
+    @property
+    def nplike(self) -> Numpy:
+        return self._numpy
+
+    @property
+    def index_nplike(self) -> Numpy:
+        return self._numpy
+
+    def __init__(self):
+        self._numpy = Numpy.instance()
+
+    def __getitem__(self, index: KernelKeyType) -> NumpyKernel:
+        return NumpyKernel(awkward_cpp.cpu_kernels.kernel[index], index)
+
+
+class CupyBackend(Backend[Any], Singleton):
+    _cupy: Cupy
+
+    @property
+    def nplike(self) -> Cupy:
+        return self._cupy
+
+    @property
+    def index_nplike(self) -> Cupy:
+        return self._cupy
+
+    def __init__(self):
+        self._cupy = Cupy.instance()
+
+    def __getitem__(self, index: KernelKeyType) -> CupyKernel | NumpyKernel:
+        from awkward._connect import cuda
+
+        cupy = cuda.import_cupy("Awkward Arrays with CUDA")
+        _cuda_kernels = cuda.initialize_cuda_kernels(cupy)
+        func = _cuda_kernels[index]
+        if func is not None:
+            return CupyKernel(func, index)
+        return NumpyKernel(awkward_cpp.cpu_kernels.kernel[index], index)
```

Comment:
@jpivarski do you know why we have this fallback? I imagine it would be proper just to fail here for missing kernels.

Reply:
Let's remove it. It hides cases in which the CuPy kernel doesn't exist, making it look like something works in the `"cuda"` backend when really, it's just copying the array to the CPU.

Thanks for finding it.

```suggestion
        else:
            raise ak._errors.wrap_error(AssertionError(f"CuPyKernel not found: {index!r}"))
```